### ISSUE TITLE
OptionROM not working with Dasharo OVMF while dispatching Dasharo image

### ISSUE THREAD

**Issue by maheshtammisetti** [Dasharo/dasharo-issues#399]
**Dasharo version**

Dasharo (UEFI) v0.1.0 for QEMU Q35

**Dasharo variant**

QEMU Q35 variant

**Affected component(s) or functionality**

While booting Dasharo OVMF with OptionROM enabled, the 3rd party image load assertion is triggered.

**Brief summary**

Once the Dasharo DEBUG OVMF image is built, starting it with QEMU Q35 and OptionROM enabled triggers the bug and stops the booting at the triggered assertion mentioned in the image. 

**How reproducible**

100%

**How to reproduce**

1.  Build the Dasharo OVMF in Docker using the following command - `build -p OvmfPkg/OvmfPkgX64.dsc -b DEBUG -t GCC5 -a X64 -n $(nproc)`
2. Boot the Dasharo OVMF with QEMU Q35 with the following command - `qemu-system-x86_64 -machine q35,smm=on -global driver=cfi.pflash01,property=secure,value=on -drive if=pflash,format=raw,unit=0,file=Build/OvmfX64/DEBUG_GCC5/FV/OVMF_CODE.fd,readonly=on -drive if=pflash,format=raw,unit=1,file=Build/OvmfX64/DEBUG_GCC5/FV/OVMF_VARS.fd -debugcon file:debug.log -global isa-debugcon.iobase=0x402 -global ICH9-LPC.disable_s3=1`

**Expected behavior**

Dasharo OVMF was supposed to boot with the OptionROM enabled, and enter the Setup Menu.

**Actual behavior**

Booting halts with the assertion triggered and Dasharo OVMF doesn't boot to Setup Menu with OptionROM enabled.

**Screenshots**

![Screenshot from 2023-03-30 00-09-28](https://user-images.githubusercontent.com/119837581/228636506-85ced425-2814-41da-919f-e0e4261af518.png)

**Additional context**

EDK2 images execute with the OptionROM enabled, but for Dasharo OVMF the 3rd party image assertion is being triggered. 

**Solutions you've tried**

Tried to analyze the Defer3rdPartyImageLoad.c file for potential workarounds, but couldn't find any.

**Comment by maheshtammisetti** [comment]
1.) FileFromFv Function - Checks the file whether it is from Firmware volume or not by calling the global Boot Services and calls the protocol LocateDevicePath(). 

if EFI_ERROR == Success then gBS->OpenProtocol() is called to know whether the handle supports a protocol or not. EFI_OPEN_PROTOCOL_TEST_PROTOCOL is passed as an attribute to test whether the protocol 

exists on a handle or not.


2.) LookupImage Function - Finds the image matching with the device path, if the deferred image is found from the image table then the memory address of the found image is returned to the function as 

pointer to the function(). Index & DeviceSize variables are used as local variables to for loop counter and to save the size of the image in order to use it in the CopyMem() function.


3.) QueueImage Function - Adds the found image to the Image Table by reallocating some space. Saves the deferred image devicepath in the table with ImageInfo->DevicePath , checks the Image->Bootoption 

whether TRUE or FALSE. Sets the Image->Loaded to FALSE, as the image is not loaded to memory instead being saved to the image table, increments the counter by 1.


4.) GetDeferredImageInfo Function - Information about the deferred image is returned, and information about all the deferred images is returned until the Index reaches EFI_NOT_FOUND. If the deferred 

images from the list are not loaded then the deferred images are moved to the new table with NewIndex which is also known that if the deferred images are loaded then they are removed. Get the deferred 

image information stored in ImageDevicePath & BootOption.


5.) EndofDxeEvent Function - Notifies the end of the Dxe before invoking any 3rd party UEFI applications or UEFI drivers.


6.) DxeSmmReadyToLock Function - The DxeSmmReadyToLock protocol is installed right after the EndOfDxe function in order to lock or protect the resources. Takes Event, Context as parameters where Event 

is the event being processed. Uses gBS->LocateProtocol() in order to find the first handle that supports the protocol in this case "gEfiDxeSmmReadyToLockProtocol". And returns status as success if the 

matching protocol is found and closes the Event (EndOfDxe). IF mImageLoadedAfterEndOfDxe then assertion is triggered to dispatch the images only after installing DxeSmmReadyToLock protocol.


7.) Defer3rdPartyImageLoad - Declares DEFERRED_3RD_PARTY_IMAGE_INFO structure to a pointer *ImageInfo. 

// if condition ignores if there is no file found. 

// if condition calls the function if(FileFromFv(File)) states if the file is from the firmware volume then returns EFI_SUCCESS.

Calls the function() LookupImage(File, BootPolicy) checks with the File(indicates file path) &  BootPolicy that the request is originated from the boot manager and the file path is being loaded by the

boot manager. Stores the device path information and prints the DEBUG Information. 

// checks that if the devicepath actually exists and != NULL then FreePool(DevicePath).

// checks if(mEndOfDxe) is actually TRUE then the image is actually loaded after the end of dxe. 

// checks whether the ImageInfo is NULL or not, if the ImageInfo is not NULL then the ImageInfo->Loaded is set to TRUE indicating it is loaded before end of dxe. 

// if(ImageInfo == NULL) QueueImage(File, BootPolicy) is called and if the Image is being loaded second time after EndOfDxe event EFI_ACCESS_DENIED is returned as ImageInfo will not be equal to NULL on 

second load.

8.) Defer3rdPartyImageLoadInitialize - Set Handle to NULL in the creation of new handle, call the gBS->InstallMultipleProtocolInterfaces() which creates a new handle stores it in the handle database

 and installs the protocol on the newly created handle in this case the protocol is "gefiDeferredImageLoadProtocol" and the interface type is passed which is "mDeferredImageLoad", which has the function

"GetDeferredImageInfo". An assertion is triggered if the status is not SUCCESS. After installing the "DeferredImageLoadProtocol" its listens to the "EndOfDxe".

gBS->CreateEventEx() Creates an event that indicates the "EndOfDxe". An assertion is triggered if the status is not SUCCESS. 

EfiCreateProtocolNotifyEvent function is called when the protocol DxeSmmReadyToLock is installed, "&gEfiDxeSmmReadyToLock" GUID is passed as an argument to the function.

**Comment by miczyg1** [comment]
This is a nice stacktrace. @maheshtammisetti could you summarize it like in one or two sentences what was the cause and the fix?

**Comment by maheshtammisetti** [comment]
The issue was caused due to the change in flag of EndofDxe in Defer3rdPartyImageLoad.c - https://github.com/Dasharo/edk2/commit/976412e7dd778ce281c5a138cb00dbe54a293e67.

And the fix would be to revert the value to TRUE from FALSE - https://github.com/Dasharo/edk2/pull/44#discussion_r1168767969
